### PR TITLE
feat: 🎸 Added majority command to get majority vote percentages; Added anti Discord invite feature

### PR DESCRIPTION
**Pull request by Snaacky**
This PR implements a majority/super-majority command that will return the amount of votes needed for a vote to pass with a majority (60%) vote and a super-majority (80%) vote based on the amount of users that have the staff role configured under settings.

Commands:
`=majority`
`=maj (alias)`

This PR implements an anti Discord invite feature which will automatically remove Discord invites sent by users without the staff role outside of channels on the whitelist. The syntax is relatively similar to the AutoMod syntax. 

Commands:
`=settings antiinvite true/false`
`=antiinvite`
`=ai (alias)`
`=antiinvite add 553311497279897601`
`=antiinvite list`
`=antiinvite remove 1`

RethinkDB insert queries:
```r.db("setsudo").table("settings").update({"antiinvite": false})```
```r.db("setsudo").table("settings").update({"antiinvitewhitelist": []})```

### CODE REVIEW

File: src/Constants.js [hunk 1]
```diff
@@ -26,6 +26,9 @@ module.exports = {
     "delete": "0xFF0000"
   },
   "defaultSettings": {
+    "antiinvite": null, // Boolean
```
Comment (SirPacker) [inline src/Constants.js]: Booleans should be false by default

File: src/Constants.js [hunk 2]
```diff
@@ -26,6 +26,9 @@ module.exports = {
...
+    "antiinvitewhitelist": null, // Array of strings
```
Comment (SirPacker) [inline src/Constants.js]: Arrays of strings should be empty arrays by default

File: src/commands/public/settings.js [hunk 3]
```diff
@@ -108,4 +110,25 @@ module.exports = class extends Command {
...
+   * 
+   * @param {Message} message The message that invoked the command
+   * @param {object} gSettings The guild's settings
+   * @param {string} setting The setting to be changed
+   * @param {string} value The boolean to be stored
+   * 
+   * @returns {string} Output to be returned to user
+   */
+  _setBoolean(message, gSettings, setting, value) {
+    if (value != "true" && value != "false") return `The provided value is not a boolean: ${value}`;
```
Comment (SirPacker) [inline src/commands/public/settings.js]: This would be better handled by regex

File: src/commands/public/settings.js [hunk 4]
```diff
@@ -38,8 +38,10 @@ module.exports = class extends Command {
     // If both a setting and value is provided
     if (match[1] && match[2]) {
       match[1] = match[1].toLowerCase();
+      match[2] = match[2].toLowerCase();
```
Comment (SirPacker) [inline src/commands/public/settings.js]: This change may cause issues with detecting role and channel names

File: src/commands/smashbros/majority.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+          {
+            "name": "Majority:",
+            "value": `${Math.round((modCount / 100) * 60 )}`,
+          },
+          {
+            "name": "Super-majority:",
+            "value": `${Math.round((modCount / 100) * 80 )}`,
+          }
+        ]
+      };
```
Comment (SirPacker) [inline src/commands/smashbros/majority.js]: Discord.js' `MessageEmbed` class should be used to construct embeds. Check other commands and the D.js documentation for how to use it.

File: src/events/message.js [hunk 6]
```diff
@@ -23,13 +24,16 @@ module.exports = class extends Event {
...
-      
+
       await this.client.handlers.db.insert("settings", defaultSettings);
     }
 
     // Check for automod violations
     await this.autoModCheck(ctx);
 
+    // Check for Discord invites
+    if (gSettings["antiinvite"] == "true") await this.antiInviteCheck(ctx);
```
Comment (SirPacker) [inline src/events/message.js]: The setting should be a boolean rather than a string

File: src/events/message.js [hunk 7]
```diff
@@ -106,13 +110,69 @@ module.exports = class extends Event {
...
+    // Check if guild has any channels in its whitelist
+    if (gSettings["antiinvitewhitelist"] && gSettings["antiinvitewhitelist"].length) {
+
+      // Loop over each whitelist entry
+      for (const channel of gSettings["antiinvitewhitelist"]) {
+        
+        if (message.channel.id != channel) {
+
+          // Construct and test regex to search for Discord invites
+          const checkRegex = new RegExp(`(https:\/\/)?(www\.)?(?:discord\.(?:gg|io|me|li)|discordapp\.com\/invite)\/([a-z0-9-.]+)?`, "i");
```
Comment (SirPacker) [inline src/events/message.js]: No variables are used in the regex here, so constructing view `new` is unnecessary. Use `/regex/` syntax.

File: src/events/message.js [hunk 8]
```diff
@@ -106,13 +110,69 @@ module.exports = class extends Event {
...
+    // Allow the message to send if the sender was a staff member
+    if (gSettings["staffrole"] && message.member.roles.some(r => r.id === gSettings["staffrole"])) return;
+
+    // Check if guild has any channels in its whitelist
+    if (gSettings["antiinvitewhitelist"] && gSettings["antiinvitewhitelist"].length) {
+
+      // Loop over each whitelist entry
+      for (const channel of gSettings["antiinvitewhitelist"]) {
+        
+        if (message.channel.id != channel) {
```
Comment (SirPacker) [inline src/events/message.js]: This should be a return check on positive rather than execution on negative

File: src/handlers/antiInvite.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+      let toSend = `__**Anti-invite whitelist for ${message.guild.name}**__\`\`\`md\n`;
+
+      // Check if guild has whitelisted any channels
+      if (!gSettings["antiinvitewhitelist"].length) return resolve(`${message.guild.name}'s anti-invite whitelist is empty.`);
+      else {
+        // Loop over each channel
+        gSettings["antiinvitewhitelist"].forEach((val, index) => {
+          // Append channel to string in a formatted form
+          toSend += `${index + 1}. ${val}${gSettings["antiinvitewhitelist"].length > index ? `\n` : ""}`;
+          // toSend += `${index + 1}. ${val}${gSettings["automodlist"].length > index ? "\n" : ""}`;
```
Comment (SirPacker) [inline src/handlers/antiInvite.js]: This comment can be removed

**Review by a1icja — CHANGES_REQUESTED** [review]
Most of the code here is good, but there are a few things that are inconsistent or out-of-line with how the rest of the bot is designed.

File: src/commands/public/settings.js [hunk 10]
```diff
@@ -124,8 +123,11 @@ module.exports = class extends Command {
...
 
+    // Converts the string value to a boolean for storage
+    let valueAsBool = (value == 'true')
+    
     // Update guild settings
-    gSettings[setting] = value;
+    gSettings[setting] = valueAsBool;
     this.client.handlers.db.update("settings", message.guild.id, gSettings);
 
     return `Setting \`${setting}\` set to \`${value}\``;
```
Comment (SirPacker) [inline src/commands/public/settings.js]: `value` should be changed to `valueAsBool`

File: src/events/message.js [hunk 11]
```diff
@@ -126,37 +126,36 @@ module.exports = class extends Event {
...
+        if (message.channel.id === channel) return;
 
-          // Construct and test regex to search for Discord invites
-          const checkRegex = new RegExp(`(https:\/\/)?(www\.)?(?:discord\.(?:gg|io|me|li)|discordapp\.com\/invite)\/([a-z0-9-.]+)?`, "i");
-          if (checkRegex.test(message.content)) {
-
-            // Fetch messages near the violation for context
-            let nearMsgs = await message.channel.messages.fetch({ limit: 5 });
+        // Construct and test regex to search for Discord invites
+        const checkRegex = /(https:\/\/)?(www\.)?(?:discord\.(?:gg|io|me|li)|discordapp\.com\/invite)\/([a-z0-9-.]+)?/g;
```
Comment (SirPacker) [inline src/events/message.js]: The flag `i` was lost here. Should be appended to the `g` flag.

**Review by a1icja — CHANGES_REQUESTED** [review]

**Review by a1icja — APPROVED** [review]
Minor nitpicks can be refactored later. LGTM!

**Pull request merged**